### DISCUSSION TITLE
Calling Dispose on enumerator

### DISCUSSION THREAD

**Discussion by jm-merle** [sebastienros/jint#1285] [Ideas]
In C#, when your write an enumerator like so :
```
        public IEnumerable<int> Ints
        {
            get
            {
                try
                {
                      // yield return something
                }
                finally
                {
                    Trace.WriteLine("End iteration");
                }
            }
        }
```
the finally section is always executed, even if the calling code aborts the iteration by throwing exceptions. This is particularly convenient, because an iterator can be used to open a disposable resource, iterate upon it, and then be assured that the resource will always be correctly disposed.

However, when the iterator is exposed to a Jint script, this pattern does not work. The following script does not trigger the execution of the finally part :
```
for (let i of ctx.Ints)
{
  if (i === 2) throw "exception";
}
```
A possible simple extension could be to override the Close method in EnumerableIterator: 
```
            public override void Close(CompletionType completion)
            {
                (_enumerator as IDisposable)?.Dispose();
                base.Close(completion);
            }
```

**Comment by lahma** [comment]
Sounds like a logical addition to interop, would you like to create a PR with associated test case?
